tests/unit/runners/test_validate.py

```diff
@@ -241,6 +241,131 @@ def test_storage_too_few(self, mock_localclient, mock_deepsea):
+    def test_check_available_succeeds(self, mock_localclient):
+        fake_data = {'admin.ceph': {'ceph-common': {'version': '13.0.1'}},
+                     'data.ceph': {'ceph-common': {'version': '13.0.1'}}}
+
+        local = mock_localclient.return_value
+        local.cmd.return_value = fake_data
+        validator = validate.Validate("setup")
+
+        validator.uninstalled = ['admin.ceph', 'data.ceph']
+        validator._check_available()
+        assert 'ceph_version' not in validator.errors
+
+    @patch('salt.client.LocalClient')
+    def test_check_available_succeeds_with_no_minions(self, mock_localclient):
+        validator = validate.Validate("setup")
+
+        validator.uninstalled = []
+        validator._check_available()
+        assert 'ceph_version' not in validator.errors
+
+    @patch('salt.client.LocalClient')
+    def test_check_available_is_older(self, mock_localclient):
+        fake_data = {'admin.ceph': {'ceph-common': {'version': '10.1.1'}},
+                     'data.ceph': {'ceph-common': {'version': '13.0.1'}}}
+
+        local = mock_localclient.return_value
+        local.cmd.return_value = fake_data
+        validator = validate.Validate("setup")
+
+        validator.uninstalled = ['admin.ceph']
+        validator._check_available()
+        assert 'admin.ceph' in validator.errors['ceph_version']
+
+    @patch('salt.client.LocalClient')
+    def test_check_available_has_broken_version(self, mock_localclient):
+        fake_data = {'admin.ceph': {'ceph-common': {'nope': 'x.x.x'}},
+                     'data.ceph': {'ceph-common': {'version': '13.0.1'}}}
+
+        local = mock_localclient.return_value
+        local.cmd.return_value = fake_data
+        validator = validate.Validate("setup")
+
+        validator.uninstalled = ['admin.ceph']
+        validator._check_available()
+        assert 'admin.ceph' in validator.errors['ceph_version']
+
+    @patch('salt.client.LocalClient')
+    def test_check_available_has_no_repo(self, mock_localclient):
+        fake_data = {'admin.ceph': '',
+                     'data.ceph': {'ceph-common': {'version': '13.0.1'}}}
+
+        local = mock_localclient.return_value
+        local.cmd.return_value = fake_data
+        validator = validate.Validate("setup")
+
+        validator.uninstalled = ['admin.ceph']
+        validator._check_available()
+        assert 'admin.ceph' in validator.errors['ceph_version']
+
+    @patch('validate.DeepseaMinions')
+    @patch('salt.client.LocalClient')
+    def test_salt_version(self, mock_localclient, mock_deepsea):
+        fake_data = { 'admin.ceph': '2018.1.99',
+                      'data.ceph': '2018.1.99'}
+
+        local = mock_localclient.return_value
+        local.cmd.return_value = fake_data
+        validator = validate.Validate("setup", search_grains=True)
+
+        assert len(validator.passed) == 0
+        validator.salt_version()
+        assert validator.passed['salt_version'] == 'valid'
```

Comment:
could we add some tests for real-life version strings? For example, as per the commit message in 97bd8dd0765d854e1505648da1b1de36260a5f8e the version string might be " Zypper command failure: Repository 'SLE-Module-Server-Applications15-Updates' is invalid.\n[Server_Applications_Module_15_x86_64:SLE-Module-Server-Applications15-Updates|https://updates.suse.com/SUSE/Updates/SLE-Module-Server-Applications/15/x86_64/update?12G6EXPLytchxzDtQtrGRUsbLPHhWuDTz1o4dmfmH1MWnPq9oaPxPfuF1GDhPkRpznZvhxzOOI1Lt_LbJdQUy4yGwg75T_cW0WxOycIFoXCObDXyFHhwG7kr7AC27DwZV2gpKwz-ezerz3GbqCboZbpywl-NnXRSKHoi] Valid metadata not found at specified URL\nPlease check if the URIs defined for this repository are pointing to a valid repository.\nSkipping repository 'SLE-Module-Server-Applications15"

Downstream ceph RPM packages have version strings like "13.0.2.989+g371e07763c-lp150"

Reply:
The two tests for strings are test_check_installed_is_not_installed and test_check_available_has_no_repo.  The contents of the string do not change the test.  The test is for the missing key.  Neither the tests nor the checks do anything with the strings themselves.

I can update the second so that it's similar to the first.

Reply:
Correction, I just check pkg.info_available and it returns an empty string for no package in a repo.

Reply:
In my version (which is being replaced here), I have code that sanity-checks the version string returned by pkg.info_installed/pkg.info_available. 

The problem is that these Salt modules appear to assume that zypper will return the output they expect - i.e. a version-like string. If zypper returns an error, they return the text of the error as the version string. That's why I put that sanity-checking code in.